In [1]:
!pip install -q gdown
import gdown

file_id = "1LcwiIopCKRoGv_NlZRmjIeMGSmR9-bqF"
gdown.download(f"https://drive.google.com/uc?id={file_id}", output="Fgsm.zip", quiet=False)
!unzip Fgsm.zip -d fgsm_folder

Downloading...
From (original): https://drive.google.com/uc?id=1LcwiIopCKRoGv_NlZRmjIeMGSmR9-bqF
From (redirected): https://drive.google.com/uc?id=1LcwiIopCKRoGv_NlZRmjIeMGSmR9-bqF&confirm=t&uuid=c52fd07a-659c-4516-a243-f69afb120477
To: /content/Fgsm.zip
100%|██████████| 177M/177M [00:04<00:00, 41.7MB/s]


Archive:  Fgsm.zip
   creating: fgsm_folder/adversarial_images_FGSM_train/
   creating: fgsm_folder/adversarial_images_FGSM_train/eps_0.007/
  inflating: fgsm_folder/adversarial_images_FGSM_train/eps_0.007/adv_0.png  
  inflating: fgsm_folder/adversarial_images_FGSM_train/eps_0.007/adv_1.png  
  inflating: fgsm_folder/adversarial_images_FGSM_train/eps_0.007/adv_2.png  
  inflating: fgsm_folder/adversarial_images_FGSM_train/eps_0.007/adv_3.png  
  inflating: fgsm_folder/adversarial_images_FGSM_train/eps_0.007/adv_4.png  
  inflating: fgsm_folder/adversarial_images_FGSM_train/eps_0.007/adv_5.png  
  inflating: fgsm_folder/adversarial_images_FGSM_train/eps_0.007/adv_6.png  
  inflating: fgsm_folder/adversarial_images_FGSM_train/eps_0.007/adv_7.png  
  inflating: fgsm_folder/adversarial_images_FGSM_train/eps_0.007/adv_8.png  
  inflating: fgsm_folder/adversarial_images_FGSM_train/eps_0.007/adv_9.png  
  inflating: fgsm_folder/adversarial_images_FGSM_train/eps_0.007/adv_10.png  
  inflating

In [2]:
!pip install -q gdown
import gdown

# New file ID extracted from the link
file_id = "1fsgH0KBLOlYf18r51qmqF1bBbSaC4DIc"
gdown.download(f"https://drive.google.com/uc?id={file_id}", output="pgd.zip", quiet=False)
!unzip pgd.zip -d pgd_folder

Downloading...
From (original): https://drive.google.com/uc?id=1fsgH0KBLOlYf18r51qmqF1bBbSaC4DIc
From (redirected): https://drive.google.com/uc?id=1fsgH0KBLOlYf18r51qmqF1bBbSaC4DIc&confirm=t&uuid=94888b1b-39e2-4f9e-aea2-8b18673ae76f
To: /content/pgd.zip
100%|██████████| 203M/203M [00:04<00:00, 50.3MB/s]


Archive:  pgd.zip
   creating: pgd_folder/adversarial_images_PGD_train/
   creating: pgd_folder/adversarial_images_PGD_train/eps_0.007/
  inflating: pgd_folder/adversarial_images_PGD_train/eps_0.007/adv_0.png  
  inflating: pgd_folder/adversarial_images_PGD_train/eps_0.007/adv_1.png  
  inflating: pgd_folder/adversarial_images_PGD_train/eps_0.007/adv_2.png  
  inflating: pgd_folder/adversarial_images_PGD_train/eps_0.007/adv_3.png  
  inflating: pgd_folder/adversarial_images_PGD_train/eps_0.007/adv_4.png  
  inflating: pgd_folder/adversarial_images_PGD_train/eps_0.007/adv_5.png  
  inflating: pgd_folder/adversarial_images_PGD_train/eps_0.007/adv_6.png  
  inflating: pgd_folder/adversarial_images_PGD_train/eps_0.007/adv_7.png  
  inflating: pgd_folder/adversarial_images_PGD_train/eps_0.007/adv_8.png  
  inflating: pgd_folder/adversarial_images_PGD_train/eps_0.007/adv_9.png  
  inflating: pgd_folder/adversarial_images_PGD_train/eps_0.007/adv_10.png  
  inflating: pgd_folder/adversarial_im

In [3]:
import os
import shutil
import pandas as pd

# Paths
root_folder = '/content/fgsm_folder/adversarial_images_FGSM_train'
csv_path = os.path.join(root_folder, 'adversarial_labels.csv')
output_folder = os.path.join(root_folder, 'fgsm_all_adversarial_images')

# Create output folder if it doesn't exist
os.makedirs(output_folder, exist_ok=True)

# Read the CSV file
df = pd.read_csv(csv_path)

# Process each row in the CSV
for _, row in df.iterrows():
    src_path = os.path.join(root_folder, row['filename'])
    dst_filename = f"adv_{row['index']}_eps{row['epsilon']}.png"  # New filename includes epsilon
    dst_path = os.path.join(output_folder, dst_filename)

    # Copy the image
    shutil.copy(src_path, dst_path)

    # Optional: You could also create a new CSV with the new filenames
    # if you need to maintain the label mapping

print(f"All adversarial images have been copied to {output_folder}")

All adversarial images have been copied to /content/fgsm_folder/adversarial_images_FGSM_train/fgsm_all_adversarial_images


In [4]:
import os
import shutil
import pandas as pd

# Paths
root_folder = '/content/fgsm_folder/adversarial_images_FGSM_train'
csv_path = os.path.join(root_folder, 'adversarial_labels.csv')
output_folder = os.path.join('/content/All_fgsm', 'all_adversarial_images')

# Create output folder if it doesn't exist
os.makedirs(output_folder, exist_ok=True)

# Read the CSV file
df = pd.read_csv(csv_path)

# Process each row in the CSV
for _, row in df.iterrows():
    src_path = os.path.join(root_folder, row['filename'])
    dst_filename = f"label_{row['label']}_eps{row['epsilon']}_idx{row['index']}.png"  # Example: "label_0_eps0.007_idx0.png"
    dst_path = os.path.join(output_folder, dst_filename)

    # Copy the image
    shutil.copy(src_path, dst_path)

print(f"All adversarial images have been copied to {output_folder} with label-based names.")

All adversarial images have been copied to /content/All_fgsm/all_adversarial_images with label-based names.


In [5]:
output_folder = os.path.join('/content/All_fgsm', 'all_adversarial_images')
# Count the number of files in the output folder
num_files = len([f for f in os.listdir(output_folder) if os.path.isfile(os.path.join(output_folder, f))])
print(f"Total number of files in {output_folder}: {num_files}")

Total number of files in /content/All_fgsm/all_adversarial_images: 2064


In [6]:
import os
import shutil
import pandas as pd

# Paths
root_folder = '/content/pgd_folder/adversarial_images_PGD_train'
csv_path = os.path.join(root_folder, 'adversarial_labels.csv')
output_folder = os.path.join('/content/All_pgd', 'pgd_all_adversarial_images')

# Create output folder if it doesn't exist
os.makedirs(output_folder, exist_ok=True)

# Read the CSV file
df = pd.read_csv(csv_path)

# Process each row in the CSV
for _, row in df.iterrows():
    src_path = os.path.join(root_folder, row['filename'])
    dst_filename = f"label_{row['label']}_eps{row['epsilon']}_idx{row['index']}.png"  # Example: "label_0_eps0.007_idx0.png"
    dst_path = os.path.join(output_folder, dst_filename)

    # Copy the image
    shutil.copy(src_path, dst_path)

print(f"All adversarial images have been copied to {output_folder} with label-based names.")

All adversarial images have been copied to /content/All_pgd/pgd_all_adversarial_images with label-based names.


In [7]:
output_folder2 = os.path.join('/content/All_pgd', 'pgd_all_adversarial_images')
# Count the number of files in the output folder
num_files = len([f for f in os.listdir(output_folder2) if os.path.isfile(os.path.join(output_folder2, f))])
print(f"Total number of files in {output_folder2}: {num_files}")

Total number of files in /content/All_pgd/pgd_all_adversarial_images: 2064


In [9]:
import os
import shutil

# Define paths (adjust these based on your actual directory structure)
fgsm_folder = "/content/All_fgsm/all_adversarial_images"  # Path to FGSM adversarial images
pgd_folder = "/content/All_pgd/pgd_all_adversarial_images"    # Path to PGD adversarial images
output_folder = "/content/All_fgsm_pgd"               # Output folder (can be anywhere)

# Create output folder if it doesn't exist
os.makedirs(output_folder, exist_ok=True)

# Function to copy files and add prefix to avoid conflicts
def copy_files(source_folder, output_folder, prefix=""):
    for filename in os.listdir(source_folder):
        src_path = os.path.join(source_folder, filename)
        if os.path.isfile(src_path):
            new_filename = f"{prefix}{filename}"
            dst_path = os.path.join(output_folder, new_filename)
            shutil.copy2(src_path, dst_path)

# Copy FGSM files with "fgsm_" prefix
copy_files(fgsm_folder, output_folder, prefix="fgsm_")

# Copy PGD files with "pgd_" prefix
copy_files(pgd_folder, output_folder, prefix="pgd_")

# Verify the total number of files combined
num_files = len(os.listdir(output_folder))
print(f"Successfully combined {num_files} files into {output_folder}.")

Successfully combined 4128 files into /content/All_fgsm_pgd.


In [10]:
# Direct download from original source (Institut für Neuroinformatik)
!wget https://sid.erda.dk/public/archives/daaeac0d7ce1152aea9b61d9f1e19370/GTSRB_Final_Training_Images.zip
#!wget https://sid.erda.dk/public/archives/daaeac0d7ce1152aea9b61d9f1e19370/GTSRB_Final_Test_Images.zip

# Unzip both files
!unzip GTSRB_Final_Training_Images.zip
#!unzip GTSRB_Final_Test_Images.zip

# Cleanup (optional)
!rm *.zip

print("Dataset downloaded and extracted!")

Streaming output truncated to the last 5000 lines.
  inflating: GTSRB/Final_Training/Images/00035/00000_00020.ppm  
  inflating: GTSRB/Final_Training/Images/00035/00000_00021.ppm  
  inflating: GTSRB/Final_Training/Images/00035/00000_00022.ppm  
  inflating: GTSRB/Final_Training/Images/00035/00000_00023.ppm  
  inflating: GTSRB/Final_Training/Images/00035/00000_00024.ppm  
  inflating: GTSRB/Final_Training/Images/00035/00000_00025.ppm  
  inflating: GTSRB/Final_Training/Images/00035/00000_00026.ppm  
  inflating: GTSRB/Final_Training/Images/00035/00000_00027.ppm  
  inflating: GTSRB/Final_Training/Images/00035/00000_00028.ppm  
  inflating: GTSRB/Final_Training/Images/00035/00000_00029.ppm  
  inflating: GTSRB/Final_Training/Images/00035/00001_00000.ppm  
  inflating: GTSRB/Final_Training/Images/00035/00001_00001.ppm  
  inflating: GTSRB/Final_Training/Images/00035/00001_00002.ppm  
  inflating: GTSRB/Final_Training/Images/00035/00001_00003.ppm  
  inflating: GTSRB/Final_Training/Image

In [11]:
import os
import shutil
from pathlib import Path

# Define paths
original_dir = 'GTSRB/Final_Training/Images'
output_dir = 'GTSRB_Flattened_Training'

# Create output directory if it doesn't exist
Path(output_dir).mkdir(parents=True, exist_ok=True)

# Process each class folder
for class_folder in sorted(os.listdir(original_dir)):
    class_path = os.path.join(original_dir, class_folder)

    # Skip if not a directory
    if not os.path.isdir(class_path):
        continue

    # Get all images in the class folder
    for image_file in os.listdir(class_path):
        if image_file.lower().endswith(('.ppm', '.png', '.jpg', '.jpeg')):
            # Create new filename with class label
            new_filename = f"class{class_folder}_{image_file}"

            # Copy (or move) the file to the new directory with new name
            shutil.copy2(
                os.path.join(class_path, image_file),
                os.path.join(output_dir, new_filename)
            )

print(f"All images have been renamed and moved to {output_dir}")

All images have been renamed and moved to GTSRB_Flattened_Training


In [12]:
import os

# Path to the flattened directory
output_dir = 'GTSRB_Flattened_Training'

# Count the number of image files
num_samples = len([f for f in os.listdir(output_dir) if f.lower().endswith(('.ppm', '.png', '.jpg', '.jpeg'))])

print(f"Total number of samples: {num_samples}")

Total number of samples: 39209


In [13]:
import os
import cv2
from tqdm import tqdm

# Define paths
original_path = 'GTSRB/Final_Training/Images'
output_path = 'GTSRB_processed'

# Create output directory if it doesn't exist
os.makedirs(output_path, exist_ok=True)

# Process each class folder
for class_folder in tqdm(sorted(os.listdir(original_path))):
    class_path = os.path.join(original_path, class_folder)

    # Skip if not a directory
    if not os.path.isdir(class_path):
        continue

    # Get all image files in the class folder
    for image_file in os.listdir(class_path):
        # Skip non-image files (like .csv files)
        if not image_file.lower().endswith(('.ppm', '.png', '.jpg', '.jpeg')):
            continue

        # Read the image
        img_path = os.path.join(class_path, image_file)
        img = cv2.imread(img_path)

        # Create new filename: label_adv_index.png
        # Remove file extension and any numbers after it (some files have .ppm.1)
        base_name = os.path.splitext(image_file)[0].split('.')[0]
        new_filename = f"{class_folder}_adv_{base_name}.png"

        # Save as PNG
        output_file = os.path.join(output_path, new_filename)
        cv2.imwrite(output_file, img)

print("Processing complete! All images have been converted to PNG and renamed.")

100%|██████████| 43/43 [00:15<00:00,  2.86it/s]

Processing complete! All images have been converted to PNG and renamed.


In [14]:
!pip install gdown
!gdown https://drive.google.com/uc?id=1hcxmSuape1ffax3Qgzmh4q2XdUGvK5mq
!unzip UAP_train10.zip -d UAP_train10
!gdown https://drive.google.com/uc?id=1SrxZyfgqlKQFUQEKqobDhOVGwKiBiUVg

Downloading...
From (original): https://drive.google.com/uc?id=1hcxmSuape1ffax3Qgzmh4q2XdUGvK5mq
From (redirected): https://drive.google.com/uc?id=1hcxmSuape1ffax3Qgzmh4q2XdUGvK5mq&confirm=t&uuid=6b996923-80dd-4a6f-98e5-90d6b0636715
To: /content/UAP_Delta10_adv_training_samples.zip
100% 233M/233M [00:05<00:00, 41.1MB/s]
unzip:  cannot find or open UAP_train10.zip, UAP_train10.zip.zip or UAP_train10.zip.ZIP.
Downloading...
From: https://drive.google.com/uc?id=1SrxZyfgqlKQFUQEKqobDhOVGwKiBiUVg
To: /content/adv_predictions_results_delta10.csv
100% 44.5k/44.5k [00:00<00:00, 76.1MB/s]


In [15]:
!unzip /content/UAP_Delta10_adv_training_samples.zip -d UAP_train10

Archive:  /content/UAP_Delta10_adv_training_samples.zip
  inflating: UAP_train10/00007_0042.png  
  inflating: UAP_train10/00003_0012.png  
  inflating: UAP_train10/00016_0043.png  
  inflating: UAP_train10/00037_0047.png  
  inflating: UAP_train10/00017_0009.png  
  inflating: UAP_train10/00019_0017.png  
  inflating: UAP_train10/00030_0032.png  
  inflating: UAP_train10/00008_0005.png  
  inflating: UAP_train10/00018_0020.png  
  inflating: UAP_train10/00025_0010.png  
  inflating: UAP_train10/00021_0031.png  
  inflating: UAP_train10/00001_0037.png  
  inflating: UAP_train10/00029_0042.png  
  inflating: UAP_train10/00013_0045.png  
  inflating: UAP_train10/00019_0008.png  
  inflating: UAP_train10/00003_0049.png  
  inflating: UAP_train10/00004_0035.png  
  inflating: UAP_train10/00037_0021.png  
  inflating: UAP_train10/00008_0048.png  
  inflating: UAP_train10/00027_0041.png  
  inflating: UAP_train10/00014_0028.png  
  inflating: UAP_train10/00023_0046.png  
  inflating: UAP_tra

In [16]:
!ls UAP_train10  # Should list adversarial images


00000_0000.png	00014_0017.png	00028_0034.png
00000_0001.png	00014_0018.png	00028_0035.png
00000_0002.png	00014_0019.png	00028_0036.png
00000_0003.png	00014_0020.png	00028_0037.png
00000_0004.png	00014_0021.png	00028_0038.png
00000_0005.png	00014_0022.png	00028_0039.png
00000_0006.png	00014_0023.png	00028_0040.png
00000_0007.png	00014_0024.png	00028_0041.png
00000_0008.png	00014_0025.png	00028_0042.png
00000_0009.png	00014_0026.png	00028_0043.png
00000_0010.png	00014_0027.png	00028_0044.png
00000_0011.png	00014_0028.png	00028_0045.png
00000_0012.png	00014_0029.png	00028_0046.png
00000_0013.png	00014_0030.png	00028_0047.png
00000_0014.png	00014_0031.png	00028_0048.png
00000_0015.png	00014_0032.png	00028_0049.png
00000_0016.png	00014_0033.png	00029_0000.png
00000_0017.png	00014_0034.png	00029_0001.png
00000_0018.png	00014_0035.png	00029_0002.png
00000_0019.png	00014_0036.png	00029_0003.png
00000_0020.png	00014_0037.png	00029_0004.png
00000_0021.png	00014_0038.png	00029_0005.png
00000_0022

In [18]:
!head '/content/UAP_train10/adv_predictions_results_delta10.csv'

filename,true_label,predicted_label
00000_0000.png,0,16
00000_0001.png,0,16
00000_0002.png,0,0
00000_0003.png,0,0
00000_0004.png,0,1
00000_0005.png,0,0
00000_0006.png,0,0
00000_0007.png,0,4
00000_0008.png,0,0


In [20]:
# Install gdown if not already installed
!pip install gdown

# Download the zip file from Google Drive
!gdown --id 1j_z95ubzDCLaZAD5JVw-PnGHDc5H1JSS

# Unzip the file into a folder named UAP_delta5
!unzip -q "UAP_Delta5_adv_training_samples.zip" -d "UAP_delta5"

# Verify the contents
!ls UAP_delta5

/usr/local/lib/python3.11/dist-packages/gdown/__main__.py:140: FutureWarning: Option `--id` was deprecated in version 4.3.1 and will be removed in 5.0. You don't need to pass it anymore to use a file ID.
  warnings.warn(
Downloading...
From (original): https://drive.google.com/uc?id=1j_z95ubzDCLaZAD5JVw-PnGHDc5H1JSS
From (redirected): https://drive.google.com/uc?id=1j_z95ubzDCLaZAD5JVw-PnGHDc5H1JSS&confirm=t&uuid=ff791d98-3802-4ed8-8079-6b42e6b3bed8
To: /content/UAP_Delta5_adv_training_samples.zip
100% 203M/203M [00:04<00:00, 43.0MB/s]
00000_0000.png	00008_0030.png	00017_0010.png	00025_0040.png	00034_0020.png
00000_0001.png	00008_0031.png	00017_0011.png	00025_0041.png	00034_0021.png
00000_0002.png	00008_0032.png	00017_0012.png	00025_0042.png	00034_0022.png
00000_0003.png	00008_0033.png	00017_0013.png	00025_0043.png	00034_0023.png
00000_0004.png	00008_0034.png	00017_0014.png	00025_0044.png	00034_0024.png
00000_0005.png	00008_0035.png	00017_0015.png	00025_0045.png	00034_0025.png
00000_00

In [21]:
# Install gdown if not already installed (skip if already done)
!pip install gdown

# Download the .csv file
!gdown --id 1HgCDncylvlSuOO_LS6BjyBfZraI6qBps

# Verify the file is downloaded (replace 'filename.csv' with the actual name if known)
!ls

/usr/local/lib/python3.11/dist-packages/gdown/__main__.py:140: FutureWarning: Option `--id` was deprecated in version 4.3.1 and will be removed in 5.0. You don't need to pass it anymore to use a file ID.
  warnings.warn(
Downloading...
From: https://drive.google.com/uc?id=1HgCDncylvlSuOO_LS6BjyBfZraI6qBps
To: /content/adv_predictions_results_delta5.csv
100% 43.7k/43.7k [00:00<00:00, 92.3MB/s]
adv_predictions_results_delta5.csv  GTSRB_processed
All_fgsm			    GTSRB_utils.py
All_fgsm_pgd			    pgd_folder
All_pgd				    sample_data
EfficientNetB1.keras		    UAP_Delta10_adv_training_samples.zip
fgsm_folder			    UAP_Delta5_adv_training_samples.zip
GTSRB				    UAP_train10
GTSRB_Flattened_Training	    UAP_train5


In [22]:
!ls UAP_train5  # Should list adversarial images

00000_0000.png	00014_0017.png	00028_0034.png
00000_0001.png	00014_0018.png	00028_0035.png
00000_0002.png	00014_0019.png	00028_0036.png
00000_0003.png	00014_0020.png	00028_0037.png
00000_0004.png	00014_0021.png	00028_0038.png
00000_0005.png	00014_0022.png	00028_0039.png
00000_0006.png	00014_0023.png	00028_0040.png
00000_0007.png	00014_0024.png	00028_0041.png
00000_0008.png	00014_0025.png	00028_0042.png
00000_0009.png	00014_0026.png	00028_0043.png
00000_0010.png	00014_0027.png	00028_0044.png
00000_0011.png	00014_0028.png	00028_0045.png
00000_0012.png	00014_0029.png	00028_0046.png
00000_0013.png	00014_0030.png	00028_0047.png
00000_0014.png	00014_0031.png	00028_0048.png
00000_0015.png	00014_0032.png	00028_0049.png
00000_0016.png	00014_0033.png	00029_0000.png
00000_0017.png	00014_0034.png	00029_0001.png
00000_0018.png	00014_0035.png	00029_0002.png
00000_0019.png	00014_0036.png	00029_0003.png
00000_0020.png	00014_0037.png	00029_0004.png
00000_0021.png	00014_0038.png	00029_0005.png
00000_0022

In [23]:
# Count .png files in UAP_train10
!find UAP_train10 -name "*.png" | wc -l

2150


In [24]:
# Count .png files in UAP_train10
!find UAP_train5 -name "*.png" | wc -l

2150


In [25]:
# Count .png files in UAP_train10
!find GTSRB_processed -name "*.png" | wc -l

39209


In [26]:
# Count .png files in UAP_train10
!find All_fgsm_pgd -name "*.png" | wc -l

4128


In [27]:
import os
import shutil

# Define your paths (replace with your actual paths)
folder1_path = '/content/UAP_train10'  # First folder path
folder2_path = '/content/UAP_train5'  # Second folder path
merged_path = '/content/UAP_merged'    # Destination folder

# Create the merged folder if it doesn't exist
os.makedirs(merged_path, exist_ok=True)

def merge_folders(source1, source2, destination):
    # First copy all files from folder1
    for filename in os.listdir(source1):
        src_file = os.path.join(source1, filename)
        dst_file = os.path.join(destination, filename)
        shutil.copy2(src_file, dst_file)

    # Then copy files from folder2, appending '_delta5' to duplicates
    for filename in os.listdir(source2):
        src_file = os.path.join(source2, filename)
        dst_file = os.path.join(destination, filename)

        # If file exists, add '_delta5' before extension
        if os.path.exists(dst_file):
            name, ext = os.path.splitext(filename)
            new_filename = f"{name}_delta5{ext}"  # e.g., "cat.jpg" → "cat_delta5.jpg"
            dst_file = os.path.join(destination, new_filename)

        shutil.copy2(src_file, dst_file)

# Merge the folders
merge_folders(folder1_path, folder2_path, merged_path)

print(f"Merged {len(os.listdir(folder1_path)) + len(os.listdir(folder2_path))} files")
print(f"Result stored in: {merged_path}")

Merged 4302 files
Result stored in: /content/UAP_merged


In [28]:

from tensorflow.keras.models import load_model
model = load_model("/content/EfficientNetB1.keras")

In [44]:
import os
import pandas as pd
from PIL import Image

def build_dataset_csv():
    data = []

    # Process All_fgsm_pgd
    for filename in os.listdir("All_fgsm_pgd"):
        if filename.endswith(".png"):
            label = int(filename.split("_")[2])
            data.append({"image_path": f"All_fgsm_pgd/{filename}", "label": label, "source": "fgsm"})

    # Process UAP_merged
    for filename in os.listdir("UAP_merged"):
        if filename.endswith(".png"):
            label = int(filename.split("_")[0])
            data.append({"image_path": f"UAP_merged/{filename}", "label": label, "source": "uap"})

    # Process GTSRB (convert .ppm to .png if needed)
    for label_dir in os.listdir("GTSRB/Final_Training/Images"):
        label = int(label_dir)
        for img_file in os.listdir(f"GTSRB/Final_Training/Images/{label_dir}"):
            if img_file.endswith(".ppm"):
                # Convert to PNG (optional)
                img_path = f"GTSRB/Final_Training/Images/{label_dir}/{img_file}"
                png_path = img_path.replace(".ppm", ".png")
                Image.open(img_path).save(png_path)  # Save as PNG
                data.append({"image_path": png_path, "label": label, "source": "gtsrb"})

    # Save to CSV
    df = pd.DataFrame(data)
    df.to_csv("dataset.csv", index=False)
    return df

build_dataset_csv()

,image_path,label,source
0,All_fgsm_pgd/pgd_label_6_eps0.03_idx470.png,6,fgsm
1,All_fgsm_pgd/pgd_label_13_eps0.01_idx69.png,13,fgsm
2,All_fgsm_pgd/fgsm_label_15_eps0.01_idx90.png,15,fgsm
3,All_fgsm_pgd/fgsm_label_27_eps0.007_idx245.png,27,fgsm
4,All_fgsm_pgd/pgd_label_8_eps0.1_idx499.png,8,fgsm
...,...,...,...
47632,GTSRB/Final_Training/Images/00008/00000_00010.png,8,gtsrb
47633,GTSRB/Final_Training/Images/00008/00036_00025.png,8,gtsrb
47634,GTSRB/Final_Training/Images/00008/00039_00000.png,8,gtsrb
47635,GTSRB/Final_Training/Images/00008/00043_00012.png,8,gtsrb


In [31]:
#model = tf.keras.models.load_model("/content/EfficientNetB1.keras")

# Check input shape (assume original is not 240x240)
print(model.input_shape)  # e.g., (None, 224, 224, 3)

(None, 240, 240, 3)


In [33]:
import tensorflow as tf
import pandas as pd
from sklearn.model_selection import train_test_split
# Load CSV (columns: image_path, label)
df = pd.read_csv("dataset.csv")
train_df, val_df = train_test_split(df, test_size=0.2, random_state=42)

# Define data pipeline (no normalization)
def load_image(path, label):
    image = tf.io.read_file(path)
    image = tf.image.decode_png(image, channels=3)  # Values in [0, 255]
    image = tf.image.resize(image, [240, 240])      # Resize to 240x240
    return image, label

# Create datasets
train_ds = tf.data.Dataset.from_tensor_slices((train_df["image_path"], train_df["label"]))
train_ds = train_ds.map(load_image).batch(32).prefetch(tf.data.AUTOTUNE)

val_ds = tf.data.Dataset.from_tensor_slices((val_df["image_path"], val_df["label"]))
val_ds = val_ds.map(load_image).batch(32).prefetch(tf.data.AUTOTUNE)
model = tf.keras.models.load_model("/content/EfficientNetB1.keras")
print(model.input_shape)  # e.g., (None, 224, 224, 3)
input_layer = tf.keras.layers.Input(shape=(240, 240, 3))
x = tf.keras.layers.Resizing(model.input_shape[1], model.input_shape[2])(input_layer)  # Resize to original
x = model(x)  # Pass through original model (pixels remain [0, 255])
new_model = tf.keras.Model(inputs=input_layer, outputs=x)
new_model.compile(
    optimizer=tf.keras.optimizers.Adam(1e-4),
    loss="sparse_categorical_crossentropy",  # For integer labels
    metrics=["accuracy"]
)
history = new_model.fit(
    train_ds,
    validation_data=val_ds,
    epochs=10
)

(None, 240, 240, 3)
Epoch 1/10
1191/1191 ━━━━━━━━━━━━━━━━━━━━ 409s 254ms/step - accuracy: 0.9601 - loss: 0.3005 - val_accuracy: 0.9901 - val_loss: 0.0343
Epoch 2/10
1191/1191 ━━━━━━━━━━━━━━━━━━━━ 256s 194ms/step - accuracy: 0.9916 - loss: 0.0301 - val_accuracy: 0.9965 - val_loss: 0.0115
Epoch 3/10
1191/1191 ━━━━━━━━━━━━━━━━━━━━ 255s 188ms/step - accuracy: 0.9964 - loss: 0.0132 - val_accuracy: 0.9978 - val_loss: 0.0065
Epoch 4/10
1191/1191 ━━━━━━━━━━━━━━━━━━━━ 270s 195ms/step - accuracy: 0.9983 - loss: 0.0060 - val_accuracy: 0.9984 - val_loss: 0.0057
Epoch 5/10
1191/1191 ━━━━━━━━━━━━━━━━━━━━ 262s 195ms/step - accuracy: 0.9989 - loss: 0.0036 - val_accuracy: 0.9981 - val_loss: 0.0054
Epoch 6/10
1191/1191 ━━━━━━━━━━━━━━━━━━━━ 254s 189ms/step - accuracy: 0.9996 - loss: 0.0023 - val_accuracy: 0.9985 - val_loss: 0.0050
Epoch 7/10
1191/1191 ━━━━━━━━━━━━━━━━━━━━ 269s 195ms/step - accuracy: 0.9996 - loss: 0.0012 - val_accuracy: 0.9987 - val_loss: 0.0040
Epoch 8/10
1191/1191 ━━━━━━━━━━━━━━━━━━━━ 

In [35]:
# Save the trained model
new_model.save("Adverserially_trained_effnet.keras")  # or .h5 format


In [45]:
import os
import shutil
import pandas as pd
from PIL import Image

def build_dataset_csv():
    target_folder = "/content/All_Adv_plus_Training_samples"  # Specified target folder name
    os.makedirs(target_folder, exist_ok=True)  # Create folder if it doesn't exist

    data = []
    counter = 1  # Counter for unique filenames

    # Process All_fgsm_pgd
    for filename in os.listdir("All_fgsm_pgd"):
        if filename.endswith(".png"):
            # Create new filename and copy image
            new_filename = f"img_{counter}.png"
            shutil.copy2(f"All_fgsm_pgd/{filename}", f"{target_folder}/{new_filename}")

            label = int(filename.split("_")[2])
            data.append({
                "image_path": f"{target_folder}/{new_filename}",
                "label": label,
                "source": "fgsm",
                "original_filename": filename
            })
            counter += 1

    # Process UAP_merged
    for filename in os.listdir("UAP_merged"):
        if filename.endswith(".png"):
            # Create new filename and copy image
            new_filename = f"img_{counter}.png"
            shutil.copy2(f"UAP_merged/{filename}", f"{target_folder}/{new_filename}")

            label = int(filename.split("_")[0])
            data.append({
                "image_path": f"{target_folder}/{new_filename}",
                "label": label,
                "source": "uap",
                "original_filename": filename
            })
            counter += 1

    # Process GTSRB (convert .ppm to .png)
    for label_dir in os.listdir("GTSRB/Final_Training/Images"):
        label = int(label_dir)
        for img_file in os.listdir(f"GTSRB/Final_Training/Images/{label_dir}"):
            if img_file.endswith(".ppm"):
                # Convert to PNG and save in target folder
                new_filename = f"img_{counter}.png"
                img_path = f"GTSRB/Final_Training/Images/{label_dir}/{img_file}"
                Image.open(img_path).save(f"{target_folder}/{new_filename}")

                data.append({
                    "image_path": f"{target_folder}/{new_filename}",
                    "label": label,
                    "source": "gtsrb",
                    "original_filename": img_file
                })
                counter += 1

    # Save to CSV
    df = pd.DataFrame(data)
    df.to_csv("Adv_plus_Training_dataset.csv", index=False)
    print(f"Successfully processed {len(df)} images. All images saved")

In [46]:
# At the end of your script file, add:
if __name__ == "__main__":
    result = build_dataset_csv()


Successfully processed 47637 images. All images saved


In [53]:
import gdown
import zipfile
import os

# Google Drive file ID (extracted from the link)
file_id = '1UM9YJcpjzBlcHC80uTJLLlfUmPGfSf3o'

# Download the file using gdown
url = f'https://drive.google.com/uc?id={file_id}'
output = 'downloaded_file.zip'
gdown.download(url, output, quiet=False)

# Unzip the file
with zipfile.ZipFile(output, 'r') as zip_ref:
    zip_ref.extractall('UAP_Test10')

!ls UAP_Test10

Downloading...
From (original): https://drive.google.com/uc?id=1UM9YJcpjzBlcHC80uTJLLlfUmPGfSf3o
From (redirected): https://drive.google.com/uc?id=1UM9YJcpjzBlcHC80uTJLLlfUmPGfSf3o&confirm=t&uuid=580e4385-0dd1-4308-8d20-bd0a7ab76fd6
To: /content/downloaded_file.zip
100%|██████████| 48.5M/48.5M [00:01<00:00, 48.0MB/s]


00016.png  01731.png  03616.png  05427.png  07089.png  09179.png  11079.png
00017.png  01761.png  03649.png  05455.png  07098.png  09196.png  11135.png
00018.png  01778.png  03684.png  05466.png  07185.png  09233.png  11142.png
00044.png  01825.png  03693.png  05487.png  07192.png  09248.png  11150.png
00095.png  01853.png  03732.png  05524.png  07219.png  09281.png  11176.png
00117.png  01874.png  03744.png  05535.png  07243.png  09295.png  11193.png
00137.png  01895.png  03789.png  05578.png  07264.png  09310.png  11205.png
00140.png  01926.png  03822.png  05611.png  07284.png  09360.png  11208.png
00153.png  01935.png  03835.png  05620.png  07299.png  09367.png  11230.png
00165.png  01941.png  03871.png  05664.png  07310.png  09390.png  11343.png
00235.png  01963.png  03946.png  05687.png  07346.png  09402.png  11366.png
00239.png  02018.png  03960.png  05701.png  07367.png  09418.png  11369.png
00243.png  02029.png  03971.png  05715.png  07389.png  09441.png  11382.png
00276.png  0

In [54]:
# Google Drive file ID (extracted from the link)
csv_file_id = '1ypaTf3JrygUsXoDSgX67uCzlyedfkUmS'

# Download the CSV file
csv_url = f'https://drive.google.com/uc?id={csv_file_id}'
csv_output = 'downloaded_file.csv'  # Temporary name (you can change this)
gdown.download(csv_url, csv_output, quiet=False)

# (Optional) Rename the file if needed
final_csv_name = 'UAP_Test10_data.csv'  # Change this to your desired name
os.rename(csv_output, final_csv_name)

# (Optional) Read the CSV into a Pandas DataFrame to verify
df = pd.read_csv(final_csv_name)
print("CSV file contents (first 5 rows):")
print(df.head())

Downloading...
From: https://drive.google.com/uc?id=1ypaTf3JrygUsXoDSgX67uCzlyedfkUmS
To: /content/downloaded_file.csv
100%|██████████| 8.14k/8.14k [00:00<00:00, 24.2MB/s]

CSV file contents (first 5 rows):
    filename  true_label  predicted_label
0  00243.png           0               16
1  00778.png           0               16
2  04726.png           0               16
3  06854.png           0               16
4  02045.png           0               16


In [58]:
import pandas as pd
import tensorflow as tf
from tensorflow.keras.preprocessing.image import ImageDataGenerator

# Load the CSV file (adjust path if needed)
csv_path = '/content/UAP_Test10/UAP_Test10_data.csv'
df = pd.read_csv(csv_path)

# Create ImageDataGenerator **without normalization**
test_datagen = ImageDataGenerator()  # No rescaling here!

# Prepare test generator
test_generator = test_datagen.flow_from_dataframe(
    dataframe=df,
    directory='/content/UAP_Test10',        # Folder with images
    x_col='filename',              # Column with filenames
    y_col='true_label',            # Column with true labels
    target_size=(240, 240),        # Adjust to model's input size
    batch_size=32,
    class_mode='raw',              # For integer labels
    shuffle=False                  # Keep order aligned with CSV
)

Found 516 validated image filenames.


In [62]:
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
# Get model predictions (logits)
predictions = new_model.predict(test_generator)

# Convert predictions to class labels (integers)
y_pred_post_adverserial = tf.argmax(predictions, axis=1).numpy()

# Extract true labels from the generator
y_true = test_generator.labels


# Calculate accuracy
acc_adv = accuracy_score(y_true, y_pred_post_adverserial)
print(f"\nPost-Adversarial Accuracy: {acc_adv:.4f}")

# Classification report (precision, recall, F1)
print("\nClassification Report:")
print(classification_report(y_true, y_pred_post_adverserial))

# Confusion matrix
print("\nConfusion Matrix:")
print(confusion_matrix(y_true, y_pred_post_adverserial))

17/17 ━━━━━━━━━━━━━━━━━━━━ 3s 145ms/step

Post-Adversarial Accuracy: 0.9767

Classification Report:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        12
           1       1.00      1.00      1.00        12
           2       1.00      1.00      1.00        12
           3       1.00      1.00      1.00        12
           4       1.00      1.00      1.00        12
           5       0.73      0.92      0.81        12
           6       1.00      0.92      0.96        12
           7       1.00      0.92      0.96        12
           8       0.92      1.00      0.96        12
           9       1.00      1.00      1.00        12
          10       0.92      1.00      0.96        12
          11       1.00      1.00      1.00        12
          12       0.85      0.92      0.88        12
          13       1.00      1.00      1.00        12
          14       1.00      1.00      1.00        12
          15       1.00      1.00  

In [65]:
import gdown
import zipfile
import os

# Create the UAP_test5 directory if it doesn't exist
os.makedirs('UAP_test5', exist_ok=True)

# Download the zip file
zip_url = 'https://drive.google.com/uc?id=1dPmRyQsTnc-3nU8up5yOXPg4DHSNYMrt'
zip_output = 'UAP_test5.zip'
gdown.download(zip_url, zip_output, quiet=False)

# Unzip the file to UAP_test5 folder
with zipfile.ZipFile(zip_output, 'r') as zip_ref:
    zip_ref.extractall('UAP_test5')

# Remove the zip file after extraction
os.remove(zip_output)

# Download the CSV file and rename it to Delta5.csv
csv_url = 'https://drive.google.com/uc?id=144J_WESAiGmjchdZvnFQq-eBnGGVLFMj'
csv_output = 'Delta5.csv'
gdown.download(csv_url, csv_output, quiet=False)

print("Download and extraction complete!")
print(f"Zip contents are in: UAP_test5/")
print(f"CSV file is saved as: {csv_output}")

Downloading...
From (original): https://drive.google.com/uc?id=1dPmRyQsTnc-3nU8up5yOXPg4DHSNYMrt
From (redirected): https://drive.google.com/uc?id=1dPmRyQsTnc-3nU8up5yOXPg4DHSNYMrt&confirm=t&uuid=bdd65648-d351-4866-8466-2a42502d1aa7
To: /content/UAP_test5.zip
100%|██████████| 48.5M/48.5M [00:00<00:00, 176MB/s]
Downloading...
From: https://drive.google.com/uc?id=144J_WESAiGmjchdZvnFQq-eBnGGVLFMj
To: /content/Delta5.csv
100%|██████████| 7.93k/7.93k [00:00<00:00, 17.9MB/s]

Download and extraction complete!
Zip contents are in: UAP_test5/
CSV file is saved as: Delta5.csv


In [66]:
# (Optional) Read the CSV into a Pandas DataFrame to verify
df = pd.read_csv('/content/UAP_test5/Delta5.csv')
print("CSV file contents (first 5 rows):")
print(df.head())

CSV file contents (first 5 rows):
    filename  true_label  predicted_label
0  00243.png           0                0
1  00778.png           0                0
2  04726.png           0                0
3  06854.png           0                0
4  02045.png           0                4


In [69]:
import pandas as pd
import tensorflow as tf
from tensorflow.keras.preprocessing.image import ImageDataGenerator

# Load the CSV file (adjust path if needed)
csv_path = '/content/UAP_test5/Delta5.csv'
df = pd.read_csv(csv_path)

# Create ImageDataGenerator **without normalization**
test_datagen = ImageDataGenerator()  # No rescaling here!

# Prepare test generator
test_generator2 = test_datagen.flow_from_dataframe(
    dataframe=df,
    directory='/content/UAP_test5',        # Folder with images
    x_col='filename',              # Column with filenames
    y_col='true_label',            # Column with true labels
    target_size=(240, 240),        # Adjust to model's input size
    batch_size=32,
    class_mode='raw',              # For integer labels
    shuffle=False                  # Keep order aligned with CSV
)

Found 516 validated image filenames.


In [70]:
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
# Get model predictions (logits)
predictions_delta5 = new_model.predict(test_generator)

# Convert predictions to class labels (integers)
y_pred_post_adverserial_delta5 = tf.argmax(predictions, axis=1).numpy()

# Extract true labels from the generator
y_true_delta5 = test_generator2.labels


# Calculate accuracy
acc_adv = accuracy_score(y_true_delta5, y_pred_post_adverserial_delta5)
print(f"\nPost-Adversarial Accuracy: {acc_adv:.4f}")

# Classification report (precision, recall, F1)
print("\nClassification Report:")
print(classification_report(y_true_delta5, y_pred_post_adverserial_delta5))

# Confusion matrix
print("\nConfusion Matrix:")
print(confusion_matrix(y_true_delta5, y_pred_post_adverserial_delta5))

17/17 ━━━━━━━━━━━━━━━━━━━━ 2s 84ms/step

Post-Adversarial Accuracy: 0.9767

Classification Report:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        12
           1       1.00      1.00      1.00        12
           2       1.00      1.00      1.00        12
           3       1.00      1.00      1.00        12
           4       1.00      1.00      1.00        12
           5       0.73      0.92      0.81        12
           6       1.00      0.92      0.96        12
           7       1.00      0.92      0.96        12
           8       0.92      1.00      0.96        12
           9       1.00      1.00      1.00        12
          10       0.92      1.00      0.96        12
          11       1.00      1.00      1.00        12
          12       0.85      0.92      0.88        12
          13       1.00      1.00      1.00        12
          14       1.00      1.00      1.00        12
          15       1.00      1.00   

In [74]:
model = tf.keras.models.load_model("/content/EfficientNetB1.keras")

In [72]:
import gdown
import zipfile

url = "https://drive.google.com/uc?id=1tt6pE0OpFrykOOhRRSVYfKKSQRp8MbsB"
output_path = "Fgsm_test.zip"
gdown.download(url, output_path, quiet=False)

# Unzip
with zipfile.ZipFile(output_path, 'r') as zip_ref:
    zip_ref.extractall("Fgsm_test")

Downloading...
From (original): https://drive.google.com/uc?id=1tt6pE0OpFrykOOhRRSVYfKKSQRp8MbsB
From (redirected): https://drive.google.com/uc?id=1tt6pE0OpFrykOOhRRSVYfKKSQRp8MbsB&confirm=t&uuid=4025c75a-2ce6-46e2-a1c1-0cd448e6d5ae
To: /content/Fgsm_test.zip
100%|██████████| 175M/175M [00:03<00:00, 53.4MB/s]


In [79]:
import pandas as pd
import tensorflow as tf
from tensorflow.keras.preprocessing.image import ImageDataGenerator
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# Load the CSV file
csv_path = '/content/Fgsm_test/adversarial_images/adversarial_labels.csv'
df = pd.read_csv(csv_path)

# Initialize ImageDataGenerator (no normalization)
test_datagen = ImageDataGenerator()

In [84]:
import pandas as pd

# Path to the CSV file in Colab
file_path = '/content/Fgsm_test/adversarial_images/adversarial_labels.csv'

# Read the CSV file into a DataFrame
df = pd.read_csv(file_path)

# Filter rows for epsilon = 0.007
epsilon_value = 0.007
filtered_df = df[df['epsilon'] == epsilon_value]

# Take the first 516 rows (or any other number you specify)
num_rows = 516
result_df_epsilon_1= filtered_df.head(num_rows)

# Display the result
print(result_df_epsilon_1)

# If you want to save the filtered data to a new CSV file
result_df.to_csv('filtered_epsilon_0.007.csv', index=False)

     epsilon  index               filename  label
0      0.007      0    eps_0.007/adv_0.png     20
1      0.007      1    eps_0.007/adv_1.png     17
2      0.007      2    eps_0.007/adv_2.png     29
3      0.007      3    eps_0.007/adv_3.png      4
4      0.007      4    eps_0.007/adv_4.png      0
..       ...    ...                    ...    ...
511    0.007    511  eps_0.007/adv_511.png     25
512    0.007    512  eps_0.007/adv_512.png     37
513    0.007    513  eps_0.007/adv_513.png     23
514    0.007    514  eps_0.007/adv_514.png     22
515    0.007    515  eps_0.007/adv_515.png     14

[516 rows x 4 columns]


In [90]:
import pandas as pd
from tensorflow.keras.preprocessing.image import ImageDataGenerator

# Load and filter CSV
csv_path = '/content/Fgsm_test/adversarial_images/adversarial_labels.csv'
df = pd.read_csv(csv_path)
eps_007_df = df[df['epsilon'] == 0.007].copy()  # Use .copy() to avoid warnings

# Fix filenames (remove 'eps_0.007/' prefix)
eps_007_df['filename'] = eps_007_df['filename'].str.replace('eps_0.007/', '')

# Create generator
test_datagen = ImageDataGenerator()
test_generator_eps007 = test_datagen.flow_from_dataframe(
    dataframe=eps_007_df,
    directory='/content/Fgsm_test/adversarial_images/eps_0.007',  # Parent folder of images
    x_col='filename',
    y_col='label',
    target_size=(240, 240),
    batch_size=32,
    class_mode='raw',
    shuffle=False
)

# Proceed with evaluation...

Found 516 validated image filenames.


In [91]:
import tensorflow as tf
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# Get model predictions (logits)
predictions_eps007 = new_model.predict(test_generator_eps007)

# Convert predictions to class labels (integers)
y_pred_eps007 = tf.argmax(predictions_eps007, axis=1).numpy()

# Extract true labels from the generator
y_true_eps007 = test_generator_eps007.labels

# Calculate accuracy
acc_eps007 = accuracy_score(y_true_eps007, y_pred_eps007)
print(f"\nPost-Adversarial Accuracy (epsilon=0.007): {acc_eps007:.4f}")

# Classification report (precision, recall, F1)
print("\nClassification Report:")
print(classification_report(y_true_eps007, y_pred_eps007))

# Confusion matrix
print("\nConfusion Matrix:")
print(confusion_matrix(y_true_eps007, y_pred_eps007))

/usr/local/lib/python3.11/dist-packages/keras/src/trainers/data_adapters/py_dataset_adapter.py:121: UserWarning: Your `PyDataset` class should call `super().__init__(**kwargs)` in its constructor. `**kwargs` can include `workers`, `use_multiprocessing`, `max_queue_size`. Do not pass these arguments to `fit()`, as they will be ignored.
  self._warn_if_super_not_called()


17/17 ━━━━━━━━━━━━━━━━━━━━ 2s 101ms/step

Post-Adversarial Accuracy (epsilon=0.007): 0.9826

Classification Report:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        12
           1       1.00      1.00      1.00        12
           2       1.00      1.00      1.00        12
           3       1.00      0.92      0.96        12
           4       1.00      1.00      1.00        12
           5       0.85      0.92      0.88        12
           6       1.00      0.92      0.96        12
           7       1.00      1.00      1.00        12
           8       0.92      1.00      0.96        12
           9       1.00      1.00      1.00        12
          10       1.00      1.00      1.00        12
          11       1.00      1.00      1.00        12
          12       0.85      0.92      0.88        12
          13       1.00      1.00      1.00        12
          14       1.00      1.00      1.00        12
          15       

In [92]:
import pandas as pd
import tensorflow as tf
from tensorflow.keras.preprocessing.image import ImageDataGenerator
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# Load the CSV file
csv_path = '/content/Fgsm_test/adversarial_images/adversarial_labels.csv'
df = pd.read_csv(csv_path)

# List of epsilon values to evaluate
epsilons = [0.007, 0.01, 0.03, 0.1]
for eps in epsilons:
    print(f"\n{'='*50}\nEvaluating Adversarial Images for Epsilon = {eps}\n{'='*50}")

    # Filter CSV for the current epsilon
    eps_df = df[df['epsilon'] == eps].copy()

    # Remove the 'eps_X.XXX/' prefix from filenames (if present)
    eps_df['filename'] = eps_df['filename'].str.replace(f'eps_{eps}/', '')

    # Create ImageDataGenerator (no normalization)
    test_datagen = ImageDataGenerator()

    # Prepare test generator
    test_generator = test_datagen.flow_from_dataframe(
        dataframe=eps_df,
        directory=f'/content/Fgsm_test/adversarial_images/eps_{eps}',  # Folder for current epsilon
        x_col='filename',
        y_col='label',
        target_size=(240, 240),  # Adjust to your model's input size
        batch_size=32,
        class_mode='raw',  # For integer labels
        shuffle=False  # Keep order aligned with CSV
    )

    # Skip if no images found
    if test_generator.samples == 0:
        print(f" No images found for epsilon={eps}. Check paths or filenames.")
        continue

    # Get model predictions
    predictions = new_model.predict(test_generator)
    y_pred = tf.argmax(predictions, axis=1).numpy()
    y_true = test_generator.labels

    # Calculate metrics
    acc = accuracy_score(y_true, y_pred)
    print(f"\nPost-Adversarial Accuracy (ε={eps}): {acc:.4f}")
    print("\nClassification Report:")
    print(classification_report(y_true, y_pred))
    print("\nConfusion Matrix:")
    print(confusion_matrix(y_true, y_pred))


Evaluating Adversarial Images for Epsilon = 0.007
Found 516 validated image filenames.


/usr/local/lib/python3.11/dist-packages/keras/src/trainers/data_adapters/py_dataset_adapter.py:121: UserWarning: Your `PyDataset` class should call `super().__init__(**kwargs)` in its constructor. `**kwargs` can include `workers`, `use_multiprocessing`, `max_queue_size`. Do not pass these arguments to `fit()`, as they will be ignored.
  self._warn_if_super_not_called()


17/17 ━━━━━━━━━━━━━━━━━━━━ 3s 159ms/step

Post-Adversarial Accuracy (ε=0.007): 0.9826

Classification Report:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        12
           1       1.00      1.00      1.00        12
           2       1.00      1.00      1.00        12
           3       1.00      0.92      0.96        12
           4       1.00      1.00      1.00        12
           5       0.85      0.92      0.88        12
           6       1.00      0.92      0.96        12
           7       1.00      1.00      1.00        12
           8       0.92      1.00      0.96        12
           9       1.00      1.00      1.00        12
          10       1.00      1.00      1.00        12
          11       1.00      1.00      1.00        12
          12       0.85      0.92      0.88        12
          13       1.00      1.00      1.00        12
          14       1.00      1.00      1.00        12
          15       1.00  

/usr/local/lib/python3.11/dist-packages/keras/src/trainers/data_adapters/py_dataset_adapter.py:121: UserWarning: Your `PyDataset` class should call `super().__init__(**kwargs)` in its constructor. `**kwargs` can include `workers`, `use_multiprocessing`, `max_queue_size`. Do not pass these arguments to `fit()`, as they will be ignored.
  self._warn_if_super_not_called()


17/17 ━━━━━━━━━━━━━━━━━━━━ 2s 124ms/step

Post-Adversarial Accuracy (ε=0.01): 0.9787

Classification Report:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        12
           1       1.00      1.00      1.00        12
           2       1.00      1.00      1.00        12
           3       1.00      0.92      0.96        12
           4       1.00      1.00      1.00        12
           5       0.85      0.92      0.88        12
           6       1.00      0.92      0.96        12
           7       1.00      1.00      1.00        12
           8       0.92      1.00      0.96        12
           9       1.00      1.00      1.00        12
          10       1.00      1.00      1.00        12
          11       1.00      1.00      1.00        12
          12       0.85      0.92      0.88        12
          13       1.00      1.00      1.00        12
          14       1.00      1.00      1.00        12
          15       1.00   

/usr/local/lib/python3.11/dist-packages/keras/src/trainers/data_adapters/py_dataset_adapter.py:121: UserWarning: Your `PyDataset` class should call `super().__init__(**kwargs)` in its constructor. `**kwargs` can include `workers`, `use_multiprocessing`, `max_queue_size`. Do not pass these arguments to `fit()`, as they will be ignored.
  self._warn_if_super_not_called()


17/17 ━━━━━━━━━━━━━━━━━━━━ 2s 116ms/step

Post-Adversarial Accuracy (ε=0.03): 0.9535

Classification Report:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        12
           1       1.00      1.00      1.00        12
           2       1.00      0.83      0.91        12
           3       1.00      0.92      0.96        12
           4       0.86      1.00      0.92        12
           5       0.73      0.92      0.81        12
           6       0.92      0.92      0.92        12
           7       1.00      0.92      0.96        12
           8       0.92      0.92      0.92        12
           9       1.00      0.92      0.96        12
          10       1.00      1.00      1.00        12
          11       1.00      1.00      1.00        12
          12       0.85      0.92      0.88        12
          13       1.00      1.00      1.00        12
          14       1.00      1.00      1.00        12
          15       1.00   

/usr/local/lib/python3.11/dist-packages/keras/src/trainers/data_adapters/py_dataset_adapter.py:121: UserWarning: Your `PyDataset` class should call `super().__init__(**kwargs)` in its constructor. `**kwargs` can include `workers`, `use_multiprocessing`, `max_queue_size`. Do not pass these arguments to `fit()`, as they will be ignored.
  self._warn_if_super_not_called()


17/17 ━━━━━━━━━━━━━━━━━━━━ 2s 92ms/step

Post-Adversarial Accuracy (ε=0.1): 0.8295

Classification Report:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        12
           1       0.63      1.00      0.77        12
           2       0.83      0.42      0.56        12
           3       0.91      0.83      0.87        12
           4       0.57      0.67      0.62        12
           5       0.58      0.58      0.58        12
           6       0.69      0.75      0.72        12
           7       0.82      0.75      0.78        12
           8       0.50      0.50      0.50        12
           9       0.90      0.75      0.82        12
          10       0.58      0.92      0.71        12
          11       0.86      1.00      0.92        12
          12       0.80      0.67      0.73        12
          13       1.00      1.00      1.00        12
          14       1.00      0.75      0.86        12
          15       0.92     

In [96]:
# Install gdown if not already installed
!pip install gdown

# Download the file from Google Drive
file_id = "1avRnoqSRSIpDiHL9_JrZe-UlE3pZwmMx"
!gdown --id {file_id}

# Check the downloaded filename (replace 'filename.zip' with the actual name)
!ls

# Unzip the file into a folder named Pgd_test
!unzip -q adversarial_images_PGD.zip -d Pgd_test  # Replace 'filename.zip' with the actual name

/usr/local/lib/python3.11/dist-packages/gdown/__main__.py:140: FutureWarning: Option `--id` was deprecated in version 4.3.1 and will be removed in 5.0. You don't need to pass it anymore to use a file ID.
  warnings.warn(
Downloading...
From (original): https://drive.google.com/uc?id=1avRnoqSRSIpDiHL9_JrZe-UlE3pZwmMx
From (redirected): https://drive.google.com/uc?id=1avRnoqSRSIpDiHL9_JrZe-UlE3pZwmMx&confirm=t&uuid=8d7e5f6c-eaa5-4a71-b8cf-f548fcb77f9d
To: /content/adversarial_images_PGD.zip
100% 206M/206M [00:04<00:00, 49.3MB/s]
adversarial_images_PGD.zip		  GTSRB_Flattened_Training
Adverserially_trained_effnet.keras	  GTSRB_processed
Adv_plus_Training_dataset.csv		  GTSRB_utils.py
All_Adv_plus_Training_samples		  pgd_folder
All_Adv_plus_Training_samples_zipped.zip  Pgd_test.zip
All_fgsm				  sample_data
All_fgsm_pgd				  UAP_Delta10_adv_training_samples.zip
All_pgd					  UAP_Delta5_adv_training_samples.zip
dataset.csv				  UAP_Delta5_test_examples_v2.zip
downloaded_file.zip			  UAP_merg

In [97]:
import pandas as pd
import tensorflow as tf
from tensorflow.keras.preprocessing.image import ImageDataGenerator
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# Load the CSV file
csv_path = '/content/Pgd_test/adversarial_images_PGD/adversarial_labels.csv'
df = pd.read_csv(csv_path)

# List of epsilon values to evaluate
epsilons = [0.007, 0.01, 0.03, 0.1]
for eps in epsilons:
    print(f"\n{'='*50}\nEvaluating Adversarial Images for Epsilon = {eps}\n{'='*50}")

    # Filter CSV for the current epsilon
    eps_df = df[df['epsilon'] == eps].copy()

    # Remove the 'eps_X.XXX/' prefix from filenames (if present)
    eps_df['filename'] = eps_df['filename'].str.replace(f'eps_{eps}/', '')

    # Create ImageDataGenerator (no normalization)
    test_datagen = ImageDataGenerator()

    # Prepare test generator
    test_generator = test_datagen.flow_from_dataframe(
        dataframe=eps_df,
        directory=f'/content/Pgd_test/adversarial_images_PGD/eps_{eps}',  # Folder for current epsilon
        x_col='filename',
        y_col='label',
        target_size=(240, 240),  # Adjust to your model's input size
        batch_size=32,
        class_mode='raw',  # For integer labels
        shuffle=False  # Keep order aligned with CSV
    )

    # Skip if no images found
    if test_generator.samples == 0:
        print(f" No images found for epsilon={eps}. Check paths or filenames.")
        continue

    # Get model predictions
    predictions = new_model.predict(test_generator)
    y_pred = tf.argmax(predictions, axis=1).numpy()
    y_true = test_generator.labels

    # Calculate metrics
    acc = accuracy_score(y_true, y_pred)
    print(f"\nPost-Adversarial Accuracy (ε={eps}): {acc:.4f}")
    print("\nClassification Report:")
    print(classification_report(y_true, y_pred))
    print("\nConfusion Matrix:")
    print(confusion_matrix(y_true, y_pred))


Evaluating Adversarial Images for Epsilon = 0.007
Found 516 validated image filenames.


/usr/local/lib/python3.11/dist-packages/keras/src/trainers/data_adapters/py_dataset_adapter.py:121: UserWarning: Your `PyDataset` class should call `super().__init__(**kwargs)` in its constructor. `**kwargs` can include `workers`, `use_multiprocessing`, `max_queue_size`. Do not pass these arguments to `fit()`, as they will be ignored.
  self._warn_if_super_not_called()


17/17 ━━━━━━━━━━━━━━━━━━━━ 2s 101ms/step

Post-Adversarial Accuracy (ε=0.007): 0.8818

Classification Report:
              precision    recall  f1-score   support

           0       1.00      0.83      0.91        12
           1       0.92      1.00      0.96        12
           2       1.00      0.83      0.91        12
           3       0.79      0.92      0.85        12
           4       0.83      0.83      0.83        12
           5       0.65      0.92      0.76        12
           6       0.80      0.33      0.47        12
           7       0.79      0.92      0.85        12
           8       0.79      0.92      0.85        12
           9       0.92      0.92      0.92        12
          10       0.92      0.92      0.92        12
          11       0.79      0.92      0.85        12
          12       0.83      0.83      0.83        12
          13       0.92      1.00      0.96        12
          14       1.00      1.00      1.00        12
          15       1.00  

/usr/local/lib/python3.11/dist-packages/keras/src/trainers/data_adapters/py_dataset_adapter.py:121: UserWarning: Your `PyDataset` class should call `super().__init__(**kwargs)` in its constructor. `**kwargs` can include `workers`, `use_multiprocessing`, `max_queue_size`. Do not pass these arguments to `fit()`, as they will be ignored.
  self._warn_if_super_not_called()


17/17 ━━━━━━━━━━━━━━━━━━━━ 2s 95ms/step 

Post-Adversarial Accuracy (ε=0.01): 0.5116

Classification Report:
              precision    recall  f1-score   support

           0       0.00      0.00      0.00        12
           1       0.53      0.75      0.62        12
           2       0.45      0.42      0.43        12
           3       0.50      0.50      0.50        12
           4       0.24      0.58      0.34        12
           5       0.38      0.42      0.40        12
           6       0.33      0.17      0.22        12
           7       0.33      0.58      0.42        12
           8       0.46      0.50      0.48        12
           9       0.80      0.67      0.73        12
          10       0.91      0.83      0.87        12
          11       0.53      0.83      0.65        12
          12       0.64      0.58      0.61        12
          13       0.92      1.00      0.96        12
          14       0.90      0.75      0.82        12
          15       0.71   

/usr/local/lib/python3.11/dist-packages/keras/src/trainers/data_adapters/py_dataset_adapter.py:121: UserWarning: Your `PyDataset` class should call `super().__init__(**kwargs)` in its constructor. `**kwargs` can include `workers`, `use_multiprocessing`, `max_queue_size`. Do not pass these arguments to `fit()`, as they will be ignored.
  self._warn_if_super_not_called()


17/17 ━━━━━━━━━━━━━━━━━━━━ 2s 106ms/step

Post-Adversarial Accuracy (ε=0.03): 0.2946

Classification Report:
              precision    recall  f1-score   support

           0       0.00      0.00      0.00        12
           1       0.40      0.67      0.50        12
           2       0.27      0.33      0.30        12
           3       0.33      0.33      0.33        12
           4       0.10      0.25      0.14        12
           5       0.18      0.17      0.17        12
           6       0.14      0.08      0.11        12
           7       0.13      0.25      0.17        12
           8       0.12      0.17      0.14        12
           9       0.25      0.17      0.20        12
          10       0.77      0.83      0.80        12
          11       0.43      0.75      0.55        12
          12       0.50      0.42      0.45        12
          13       0.80      0.67      0.73        12
          14       0.83      0.42      0.56        12
          15       0.00   

/usr/local/lib/python3.11/dist-packages/keras/src/trainers/data_adapters/py_dataset_adapter.py:121: UserWarning: Your `PyDataset` class should call `super().__init__(**kwargs)` in its constructor. `**kwargs` can include `workers`, `use_multiprocessing`, `max_queue_size`. Do not pass these arguments to `fit()`, as they will be ignored.
  self._warn_if_super_not_called()


17/17 ━━━━━━━━━━━━━━━━━━━━ 2s 125ms/step

Post-Adversarial Accuracy (ε=0.1): 0.0969

Classification Report:
              precision    recall  f1-score   support

           0       0.00      0.00      0.00        12
           1       0.00      0.00      0.00        12
           2       0.00      0.00      0.00        12
           3       0.00      0.00      0.00        12
           4       0.04      0.08      0.05        12
           5       0.00      0.00      0.00        12
           6       0.11      0.08      0.10        12
           7       0.00      0.00      0.00        12
           8       0.00      0.00      0.00        12
           9       0.00      0.00      0.00        12
          10       0.36      0.42      0.38        12
          11       0.23      0.50      0.32        12
          12       0.10      0.08      0.09        12
          13       0.40      0.17      0.24        12
          14       0.38      0.25      0.30        12
          15       0.00    

/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


| Attack Method       | Epsilon (ε)      | Alpha (α)       | Pre-Adversarial Accuracy | Post-Adversarial Accuracy |
|---------------------|------------------|-----------------|--------------------------|---------------------------|
| **Original Model**  | -                | -               | 96%                      | -                         |
| **FGSM**           | 0.007            | -               | 58%                      | 98.26%                    |
| **FGSM**           | 0.01             | -               | 44%                      | 97.87%                    |
| **FGSM**           | 0.03             | -               | 14%                      | 95.35%                    |
| **FGSM**           | 0.1              | -               | 6%                       | 82.95%                    |
| **PGD**            | 0.007            | 0.00175         | 35%                      | 88.18%                    |
| **PGD**            | 0.01             | 0.0025          | 7%                       | 51.16%                    |
| **PGD**            | 0.03             | 0.0075          | 3%                       | 29.46%                    |
| **PGD**            | 0.1              | 0.025           | 0%                       | 9.69%                     |
| **UAP (delta 10)**            |    | -               | 34%                      | 97.67%                    |

In [98]:
import pandas as pd
import tensorflow as tf
from tensorflow.keras.preprocessing.image import ImageDataGenerator
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
# Load the CSV file (adjust path if needed)
csv_path = '/content/UAP_test5/Delta5.csv'
df = pd.read_csv(csv_path)

# Create ImageDataGenerator **without normalization**
test_datagen = ImageDataGenerator()  # No rescaling here!

# Prepare test generator
test_generator = test_datagen.flow_from_dataframe(
    dataframe=df,
    directory='/content/UAP_test5',        # Folder with images
    x_col='filename',              # Column with filenames
    y_col='true_label',            # Column with true labels
    target_size=(240, 240),        # Adjust to model's input size
    batch_size=32,
    class_mode='raw',              # For integer labels
    shuffle=False                  # Keep order aligned with CSV
)

# Get model predictions (logits)
predictions = new_model.predict(test_generator)

# Convert predictions to class labels (integers)
y_predicted = tf.argmax(predictions, axis=1).numpy()

# Extract true labels from the generator
y_true = test_generator.labels


# Calculate accuracy
acc_adv = accuracy_score(y_true, y_predicted )
print(f"\nPost-Adversarial Accuracy: {acc_adv:.4f}")

# Classification report (precision, recall, F1)
print("\nClassification Report:")
print(classification_report(y_true, y_predicted ))

# Confusion matrix
print("\nConfusion Matrix:")
print(confusion_matrix(y_true,y_predicted ))

Found 516 validated image filenames.


/usr/local/lib/python3.11/dist-packages/keras/src/trainers/data_adapters/py_dataset_adapter.py:121: UserWarning: Your `PyDataset` class should call `super().__init__(**kwargs)` in its constructor. `**kwargs` can include `workers`, `use_multiprocessing`, `max_queue_size`. Do not pass these arguments to `fit()`, as they will be ignored.
  self._warn_if_super_not_called()


17/17 ━━━━━━━━━━━━━━━━━━━━ 2s 92ms/step

Post-Adversarial Accuracy: 0.9767

Classification Report:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        12
           1       1.00      1.00      1.00        12
           2       1.00      1.00      1.00        12
           3       1.00      1.00      1.00        12
           4       1.00      1.00      1.00        12
           5       0.73      0.92      0.81        12
           6       1.00      0.92      0.96        12
           7       1.00      0.92      0.96        12
           8       0.92      1.00      0.96        12
           9       1.00      1.00      1.00        12
          10       0.92      1.00      0.96        12
          11       1.00      1.00      1.00        12
          12       0.85      0.92      0.88        12
          13       1.00      1.00      1.00        12
          14       1.00      1.00      1.00        12
          15       1.00      1.00   

In [99]:
import pandas as pd
import tensorflow as tf
from tensorflow.keras.preprocessing.image import ImageDataGenerator
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
# Load the CSV file (adjust path if needed)
csv_path = '/content/UAP_Test10/UAP_Test10_data.csv'
df = pd.read_csv(csv_path)

# Create ImageDataGenerator **without normalization**
test_datagen = ImageDataGenerator()  # No rescaling here!

# Prepare test generator
test_generator = test_datagen.flow_from_dataframe(
    dataframe=df,
    directory='/content/UAP_Test10',        # Folder with images
    x_col='filename',              # Column with filenames
    y_col='true_label',            # Column with true labels
    target_size=(240, 240),        # Adjust to model's input size
    batch_size=32,
    class_mode='raw',              # For integer labels
    shuffle=False                  # Keep order aligned with CSV
)

# Get model predictions (logits)
predictions = new_model.predict(test_generator)

# Convert predictions to class labels (integers)
y_predicted = tf.argmax(predictions, axis=1).numpy()

# Extract true labels from the generator
y_true = test_generator.labels


# Calculate accuracy
acc_adv = accuracy_score(y_true, y_predicted )
print(f"\nPost-Adversarial Accuracy: {acc_adv:.4f}")

# Classification report (precision, recall, F1)
print("\nClassification Report:")
print(classification_report(y_true, y_predicted ))

# Confusion matrix
print("\nConfusion Matrix:")
print(confusion_matrix(y_true,y_predicted ))

Found 516 validated image filenames.


/usr/local/lib/python3.11/dist-packages/keras/src/trainers/data_adapters/py_dataset_adapter.py:121: UserWarning: Your `PyDataset` class should call `super().__init__(**kwargs)` in its constructor. `**kwargs` can include `workers`, `use_multiprocessing`, `max_queue_size`. Do not pass these arguments to `fit()`, as they will be ignored.
  self._warn_if_super_not_called()


17/17 ━━━━━━━━━━━━━━━━━━━━ 2s 114ms/step

Post-Adversarial Accuracy: 0.9767

Classification Report:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        12
           1       1.00      1.00      1.00        12
           2       1.00      1.00      1.00        12
           3       1.00      1.00      1.00        12
           4       1.00      1.00      1.00        12
           5       0.73      0.92      0.81        12
           6       1.00      0.92      0.96        12
           7       1.00      0.92      0.96        12
           8       0.92      1.00      0.96        12
           9       1.00      1.00      1.00        12
          10       0.92      1.00      0.96        12
          11       1.00      1.00      1.00        12
          12       0.85      0.92      0.88        12
          13       1.00      1.00      1.00        12
          14       1.00      1.00      1.00        12
          15       1.00      1.00  

### Edit

In [ ]:
import tensorflow as tf
from tensorflow.keras.preprocessing.image import ImageDataGenerator
import os
from tqdm import tqdm
import numpy as np
import pickle # <-- Import pickle for saving state

# ========== State Management Functions (NEW) ==========

def save_training_state(epoch, epsilon_list, val_history, state_path="training_state.pkl"):
    """Saves the current training state to a file."""
    state = {
        'epoch': epoch,
        'epsilon_list': epsilon_list,
        'val_history': val_history,
    }
    with open(state_path, 'wb') as f:
        pickle.dump(state, f)
    print(f"\n---> Training state saved to {state_path}")

def load_training_state(state_path="training_state.pkl"):
    """Loads training state from a file if it exists."""
    if os.path.exists(state_path):
        with open(state_path, 'rb') as f:
            state = pickle.load(f)
        print(f"---> Found previous state. Resuming from epoch {state['epoch']}.")
        return state['epoch'], state['epsilon_list'], state['val_history']
    else:
        print("---> No previous state found. Starting training from scratch.")
        # Return initial default values
        return 0, [], {
            'clean_loss': [], 'clean_acc': [],
            'adv_loss': [], 'adv_acc': []
        }

# ======================================================

# ========== 1. Load GTSRB Dataset Using ImageDataGenerator ==========
# (No changes here, keeping it for context)
data_dir = "GTSRB/Final_Training/Images"
test_dir = "GTSRB/Final_Test/Images"
batch_size = 128
target_size = (240, 240)

train_datagen = ImageDataGenerator(
    rotation_range=15, width_shift_range=0.1, height_shift_range=0.1,
    zoom_range=0.2, brightness_range=[0.8, 1.2], horizontal_flip=False,
    validation_split=0.1
)
val_datagen = ImageDataGenerator(validation_split=0.1)

train_generator = train_datagen.flow_from_directory(
    data_dir, target_size=target_size, batch_size=batch_size,
    class_mode="sparse", subset="training", seed=42
)
val_generator = val_datagen.flow_from_directory(
    data_dir, target_size=target_size, batch_size=batch_size,
    class_mode="sparse", subset="validation", seed=42
)

# ========== 2. FGSM and PGD Attack Functions ==========
# (No changes here)
@tf.function
def fgsm_attack_batch(model, images, labels, epsilon):
    epsilon_scaled = epsilon * 255.0
    with tf.GradientTape() as tape:
        tape.watch(images)
        predictions = model(images, training=True)
        loss = tf.keras.losses.sparse_categorical_crossentropy(labels, predictions)
    gradients = tape.gradient(loss, images)
    adv_images = images + epsilon_scaled * tf.sign(gradients)
    return tf.clip_by_value(adv_images, 0, 255)

@tf.function
def pgd_attack_batch(model, images, labels, epsilon, alpha, iterations):
    epsilon_scaled = epsilon * 255.0
    alpha_scaled = alpha * 255.0
    adv_images = tf.identity(images)
    for _ in range(iterations):
        with tf.GradientTape() as tape:
            tape.watch(adv_images)
            predictions = model(adv_images, training=True)
            loss = tf.keras.losses.sparse_categorical_crossentropy(labels, predictions)
        gradients = tape.gradient(loss, adv_images)
        adv_images = adv_images + alpha_scaled * tf.sign(gradients)
        adv_images = tf.clip_by_value(adv_images, images - epsilon_scaled, images + epsilon_scaled)
        adv_images = tf.clip_by_value(adv_images, 0, 255)
    return adv_images

# ========== 3. Advanced Training Step ==========
# (No changes here)
@tf.function
def adaptive_train_step(model, optimizer, loss_fn, x, y, epsilon_list, alpha_factor, iterations, epoch):
    bs = tf.shape(x)[0]
    base_clean_ratio = 0.75
    final_clean_ratio = 0.5
    progress = tf.cast(epoch, tf.float32) / 35.0
    clean_ratio = base_clean_ratio - (base_clean_ratio - final_clean_ratio) * progress
    clean_portion = tf.cast(tf.math.floor(clean_ratio * tf.cast(bs, tf.float32)), tf.int32)
    adv_portion = bs - clean_portion
    x_clean, y_clean = x[:clean_portion], y[:clean_portion]
    x_adv_raw, y_adv = x[clean_portion:], y[clean_portion:]
    eps_tensor = tf.convert_to_tensor(epsilon_list, dtype=tf.float32)
    eps_idx = tf.random.uniform((), minval=0, maxval=len(epsilon_list), dtype=tf.int32)
    eps = eps_tensor[eps_idx]
    alpha = eps / (iterations * 1.5)
    use_fgsm_prob = 0.7 - (progress * 0.4)
    use_fgsm = tf.random.uniform(()) < use_fgsm_prob
    if use_fgsm:
        x_adv = fgsm_attack_batch(model, x_adv_raw, y_adv, eps)
    else:
        dynamic_iterations = tf.cast(iterations + tf.cast(progress * 5, tf.int32), tf.int32)
        x_adv = pgd_attack_batch(model, x_adv_raw, y_adv, eps, alpha, dynamic_iterations)
    x_all = tf.concat([x_clean, x_adv], axis=0)
    y_all = tf.concat([y_clean, y_adv], axis=0)
    with tf.GradientTape() as tape:
        predictions = model(x_all, training=True)
        loss = loss_fn(y_all, predictions)
    grads = tape.gradient(loss, model.trainable_variables)
    optimizer.apply_gradients(zip(grads, model.trainable_variables))
    acc = tf.keras.metrics.sparse_categorical_accuracy(y_all, predictions)
    return loss, tf.reduce_mean(acc), clean_ratio

# ========== 4. Adversarial Validation Function ==========
# (No changes here)
def adversarial_validation(model, val_generator, loss_fn, epsilon_list, alpha_factor, iterations):
    clean_loss_metric, clean_acc_metric = tf.keras.metrics.Mean(), tf.keras.metrics.SparseCategoricalAccuracy()
    adv_loss_metric, adv_acc_metric = tf.keras.metrics.Mean(), tf.keras.metrics.SparseCategoricalAccuracy()
    print("  Evaluating on clean validation data...")
    for step in tqdm(range(len(val_generator)), desc="Clean Validation", leave=False):
        x_val, y_val = val_generator[step]
        x_val = tf.cast(x_val, tf.float32)
        predictions = model(x_val, training=False)
        clean_loss_metric.update_state(loss_fn(y_val, predictions))
        clean_acc_metric.update_state(y_val, predictions)
    print("  Evaluating on adversarial validation data...")
    for step in tqdm(range(len(val_generator)), desc="Adversarial Validation", leave=False):
        x_val, y_val = val_generator[step]
        x_val = tf.cast(x_val, tf.float32)
        eps = epsilon_list[len(epsilon_list)//2]
        alpha = eps * alpha_factor
        if np.random.random() < 0.5:
            x_adv = fgsm_attack_batch(model, x_val, y_val, eps)
        else:
            x_adv = pgd_attack_batch(model, x_val, y_val, eps, alpha, iterations)
        adv_predictions = model(x_adv, training=False)
        adv_loss_metric.update_state(loss_fn(y_val, adv_predictions))
        adv_acc_metric.update_state(y_val, adv_predictions)
    return (clean_loss_metric.result().numpy(), clean_acc_metric.result().numpy(),
            adv_loss_metric.result().numpy(), adv_acc_metric.result().numpy())

# ========== 5. Train the Model with Adversarial Validation (MODIFIED) ==========

# --- Configuration ---
epochs = 35
iterations = 10
alpha_factor = 1 / iterations
# This is the BASE list of epsilons. It will not be modified.
base_epsilon_list = [0.007, 0.01, 0.03, 0.05, 0.07, 0.1, 0.15]
state_file = "training_state.pkl"

# --- Load State and Model ---
start_epoch, _, val_history = load_training_state(state_file) # Use _ for epsilon_list as we recalculate it

model_path_to_load = f"efficientnet_adv_epoch_{start_epoch}.keras"

if start_epoch > 0 and os.path.exists(model_path_to_load):
    print(f"---> Loading saved model from: {model_path_to_load}")
    model = tf.keras.models.load_model(model_path_to_load)
else:
    print("---> Loading initial base model: EfficientNetB1.keras")
    model = tf.keras.models.load_model("EfficientNetB1.keras")
    # Reset history if we are starting from scratch
    val_history = {'clean_loss': [], 'clean_acc': [], 'adv_loss': [], 'adv_acc': []}

# --- Initialize Optimizer and Loss ---
optimizer = tf.keras.optimizers.Adam(1e-4)
loss_fn = tf.keras.losses.SparseCategoricalCrossentropy()

# --- Main Training Loop (MODIFIED) ---
# The loop now starts from the loaded epoch number
for epoch in range(start_epoch, epochs):
    print(f"\nEpoch {epoch+1}/{epochs}")
    print("=" * 50)

    # Dynamic epsilon scheduling - calculated fresh each epoch from the base list
    # This prevents the bug of repeated multiplication.
    epsilon_list = [e * (1 + epoch / epochs) for e in base_epsilon_list]
    print(f"Epsilon range for this epoch: {[f'{e:.4f}' for e in epsilon_list]}")

    # Training phase
    print("Training...")
    progress_bar = tqdm(range(len(train_generator)), desc=f"Epoch {epoch+1}")
    epoch_losses, epoch_accs = [], []

    for step in progress_bar:
        x_batch, y_batch = train_generator[step]
        x_batch = tf.cast(x_batch, tf.float32)
        loss, acc, clean_ratio = adaptive_train_step(
            model, optimizer, loss_fn, x_batch, y_batch,
            epsilon_list, alpha_factor, iterations, epoch
        )
        epoch_losses.append(loss.numpy())
        epoch_accs.append(acc.numpy())
        progress_bar.set_postfix({
            "Loss": f"{loss:.4f}", "Acc": f"{acc:.4f}", "Clean%": f"{clean_ratio:.0%}"
        })

    print(f"Training Summary — Loss: {np.mean(epoch_losses):.4f}, Accuracy: {np.mean(epoch_accs):.4f}")

    # Validation phase
    print("\nValidation...")
    clean_loss, clean_acc, adv_loss, adv_acc = adversarial_validation(
        model, val_generator, loss_fn, epsilon_list, alpha_factor, iterations
    )

    val_history['clean_loss'].append(clean_loss)
    val_history['clean_acc'].append(clean_acc)
    val_history['adv_loss'].append(adv_loss)
    val_history['adv_acc'].append(adv_acc)

    print(f"\nValidation Summary:")
    print(f"  Clean     — Loss: {clean_loss:.4f}, Accuracy: {clean_acc:.4f}")
    print(f"  Adversarial — Loss: {adv_loss:.4f}, Accuracy: {adv_acc:.4f}")
    print(f"  Robustness Gap: {clean_acc - adv_acc:.4f}")

    # Save checkpoint and state every 5 epochs
    if (epoch + 1) % 3 == 0:
      model_path = f"efficientnet_adv_epoch_{epoch+1}.keras"
      model.save(model_path)
      print(f"\n---> Model checkpoint saved to {model_path}")

      # Copy to Google Drive
      drive_model_path = os.path.join(DRIVE_FOLDER, model_path)
      shutil.copy(model_path, drive_model_path)
      print(f"---> Copied to Google Drive at {drive_model_path}")
      state_path = "training_state.pkl"
      drive_state_path = os.path.join(DRIVE_FOLDER,state_path)


      # Save state for resuming
      save_training_state(epoch + 1, epsilon_list, val_history, state_file)
      shutil.copy(state_file,drive_state_path)

# Save final model
model.save("efficientnet_fgsm_pgd_robust_final.keras")
print("\nFinal model saved to efficientnet_fgsm_pgd_robust_final.keras")

# Final summary using the complete history
print("\n" + "="*60)
print("FINAL TRAINING SUMMARY (Full History)")
print("="*60)
if val_history['clean_acc']:
    print(f"Best Clean Accuracy: {max(val_history['clean_acc']):.4f}")
    print(f"Best Adversarial Accuracy: {max(val_history['adv_acc']):.4f}")
    print(f"Final Clean Accuracy: {val_history['clean_acc'][-1]:.4f}")
    print(f"Final Adversarial Accuracy: {val_history['adv_acc'][-1]:.4f}")
    print(f"Final Robustness Gap: {val_history['clean_acc'][-1] - val_history['adv_acc'][-1]:.4f}")
else:
    print("No validation history recorded.")



###  FGSM Accuracy 

| Epsilon (ε) | **Before Adversarial Training (%)** | **After Adversarial Training (%)** |
| ----------- | ----------------------------------- | ---------------------------------- |
| 0.007       | 44.96                               | 77.91                              |
| 0.010       | 34.30                               | 75.39                              |
| 0.030       | 12.79                               | 75.78                              |
| 0.100       | 9.30                                | 47.67                              |

---

###  PGD Accuracy 

| Epsilon (ε) | Alpha (α) | **Before Adversarial Training (%)** | **After Adversarial Training (%)** |
| ----------- | --------- | ----------------------------------- | ---------------------------------- |
| 0.007       | 0.00175   | 36.43                               | 68.60                              |
| 0.010       | 0.0025    | 21.71                               | 61.43                              |
| 0.030       | 0.0075    | 2.71                                | 42.83                              |
| 0.100       | 0.0250    | 0.00                                | 13.76                              |
